### PR TITLE
Enable autoconnect

### PR DESCRIPTION
**Pull request by EdDev**
Enable autoconnect on profile creation.

### CODE REVIEW

**Comment by EdDev** [comment]
There seem to be a problem enabling autoconnect: https://bugzilla.redhat.com/show_bug.cgi?id=1667874

I will try to conceive a workaround for this issue (retry or sleep).

**Comment by EdDev** [comment]
ci test please

**Comment by tyll** [comment]
another possible workaround might be to activate it without autoconnect and enable autoconnect when it is already active in the configuration.

File: libnmstate/nm/device.py [hunk 1]
```diff
@@ -138,6 +147,20 @@ def _safe_activate_async(dev, connection_id):
     )
 
 
+def _get_connection(client, connection_id):
+    conn = None
+    if connection_id:
+        conn = client.get_connection_by_id(connection_id)
```
Comment (tyll) [inline libnmstate/nm/device.py]: This looks like we should have our own client class that just has a get_connection method that works like this.

File: libnmstate/nm/device.py [hunk 2]
```diff
@@ -123,13 +121,24 @@ def _safe_activate_async(dev, connection_id):
     client = nmclient.client()
     mainloop = nmclient.mainloop()
     cancellable = mainloop.new_cancellable()
-    connection = None
-    if connection_id:
-        connection = client.get_connection_by_id(connection_id)
+
+    conn = _get_connection(client, connection_id)
```
Comment (tyll) [inline libnmstate/nm/device.py]: IMHO connection is a better variable name than conn.

File: libnmstate/nm/device.py [hunk 3]
```diff
@@ -126,6 +126,9 @@ def _safe_activate_async(dev, connection_id):
     connection = None
     if connection_id:
         connection = client.get_connection_by_id(connection_id)
+    if connection:
```
Comment (tyll) [inline libnmstate/nm/device.py]: Should there be a check that ensures that it is only called with dev or connection_id? Is it going to be called with both but the connection_id invalid?

**Comment by EdDev** [comment]
The autoconnect is supposed to resolve two issues: State persistence over boot and auto activation of the devices (like in the case of bond with dhcp).

I think that the last proposal will probably not solve the second problem.

File: libnmstate/nm/connection.py [hunk 4]
```diff
@@ -108,9 +108,9 @@ def create_setting(con_name, iface_name, iface_type):
     con_setting.props.interface_name = iface_name
     con_setting.props.uuid = str(uuid.uuid4())
     con_setting.props.type = iface_type
-    con_setting.props.autoconnect = False
```
Comment (tyll) [inline libnmstate/nm/connection.py]: The copyright header needs to be updated.

File: libnmstate/nm/device.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
 
 import logging
 
+from . import connection
```
Comment (tyll) [inline libnmstate/nm/device.py]: The copyright header needs to be updated.

File: libnmstate/nm/device.py (hunk 1 above)
Comment (tyll) [reply libnmstate/nm/device.py]: Eventiually it should be a  class that inherits from the NM client class. Moving it to a module is not an improvement IMHO. It does not need to be fixed now, though.

File: libnmstate/netapplier.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright 2018 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/netapplier.py]: Done

File: libnmstate/nm/nmclient.py [hunk 7]
```diff
@@ -68,17 +68,24 @@ def __init__(self):
...
             func, args, kwargs = action
             logging.debug('Executing NM action: func=%s', func.__name__)
             func(*args, **kwargs)
         else:
             logging.debug('NM action queue exhausted, quiting mainloop')
             self._mainloop.quit()
 
+    def execute_last_action(self):
+        func, args, kwargs = self._last_action
+        logging.debug('Executing NM (last) action: func=%s', func.__name__)
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/nm/nmclient.py]: Done

File: libnmstate/nm/device.py [hunk 8]
```diff
@@ -161,6 +161,10 @@ def _active_connection_callback(src_object, result, user_data):
             logging.debug(
                 'Connection activation canceled on %s %s: error=%s',
                 activation_type, activation_object, e)
+        elif is_connection_unavailable(e):
+            logging.warning('Connection unavailable on %s %s, retrying',
+                            activation_type, activation_object)
+            mainloop.execute_last_action()
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/nm/device.py]: We have the mainloop `timeout` to protect us from such a scenario.

**Pull request merged**